### PR TITLE
3D Tiles Sandcastle demo

### PR DESCRIPTION
**Pull request by lilleyse**
For #3177 

I took most of the code from the `Cities.html` Sandcastle demo and made it into `3D Tiles.html`. I added a dropdown to switch between data sets. Right now they are pulling from the `Specs/Data` folder, but I'm happy to copy them to the `Apps/SampleData` folder if that makes more sense.

### CODE REVIEW

**Comment by pjcozzi** [comment]
This didn't require any updates to server.js?  I would expect that we should add something like
```
'application/octet-stream' : ['b3dm', /* other extensions */],
```
to here: https://github.com/AnalyticalGraphicsInc/cesium/blob/master/server.js#L41-L46

**Comment by pjcozzi** [comment]
> Right now they are pulling from the Specs/Data folder, but I'm happy to copy them to the Apps/SampleData folder if that makes more sense.

Can you add a TODO to #3241, then we'll do this before merging to master to avoid headaches when regenerating the dataset.

File: Apps/Sandcastle/gallery/3D Tiles.html [hunk 1]
```diff
@@ -29,35 +29,57 @@
...
+var cityUrl = 'http://localhost:8080/Specs/Data/Cesium3DTiles/Tilesets/Tileset/';
+var batchedUrl = 'http://localhost:8080/Specs/Data/Cesium3DTiles/Batched/BatchedWithBatchTable/';
+var instancedUrl = 'http://localhost:8080/Specs/Data/Cesium3DTiles/Instanced/InstancedWithBatchTable/';
+var compositeUrl = 'http://localhost:8080/Specs/Data/Cesium3DTiles/Composite/Composite/';
+var pointsUrl = 'http://localhost:8080/Specs/Data/Cesium3DTiles/Points/Points/';
+var washingtonDCUrl = 'http://localhost:8002/tilesets/WashingtonDC/';
+var canaryWharfUrl = 'http://localhost:8002/tilesets/CanaryWharf/';
+var seattleUrl = 'http://localhost:8002/tilesets/Seattle/';
+
+var city;
```
Comment (pjcozzi) [inline Apps/Sandcastle/gallery/3D Tiles.html]: Minor, but can you rename `city` to `tileset`?  I started using `tileset` as the canonical variable name.

File: Apps/Sandcastle/gallery/3D Tiles.html [hunk 2]
```diff
@@ -388,11 +445,11 @@
...
     var batchTableResources = content.batchTableResources;
 
-    // Is the tile a Batched 3D Model?
-    if (content instanceof Cesium.Batched3DModel3DTileContentProvider) {
+    // Is the tile a Batched 3D Model or Instanced 3D Model?
+    if (Cesium.defined(batchTableResources)) {
         // Does it have a per-model height property?
         if (batchTableResources.hasProperty(currentPropertyName)) {
-            var length = content.batchSize;
+            var length = Cesium.defaultValue(content.batchLength, content.instancesLength);
```
Comment (pjcozzi) [inline Apps/Sandcastle/gallery/3D Tiles.html]: Nice...as I said offline today, I will unify these interfaces...starting tomorrow.

**Comment by lilleyse** [comment]
> This didn't require any updates to server.js? I would expect that we should add something like

Ah thanks for catching that, I made the change to my cleanup branch but not here.

File: Source/Scene/Cesium3DTile.js [hunk 3]
```diff
@@ -405,7 +405,7 @@ define([
 
         if (owner.debugShowContentBoundingVolume && hasContentBoundingVolume && workaround2657(tile._header.content.boundingVolume)) {
             if (!defined(tile._debugContentBoundingVolume)) {
-                tile._debugContentBoundingVolume = tile._boundingVolume.createDebugVolume(Color.BLUE);
+                tile._debugContentBoundingVolume = tile._contentBoundingVolume.createDebugVolume(Color.BLUE);
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTile.js]: Good catch!

**Comment by pjcozzi** [comment]
Also, please add a TODO to #3241 to integrate the [sample CC3D data](http://cesiumjs.org/2015/04/27/3D-Buildings-in-Cesium/) into this Sandcastle example and to generate the thumbnail.

**Comment by pjcozzi** [comment]
Any idea why this isn't 11.3?  Is this a new issue with Cesium labels?  Or something in this branch?

image

```javascript
var cartesian = scene.pickPosition(movement.position);
var cartographic = Cesium.Cartographic.fromCartesian(cartesian);
var height = cartographic.height.toFixed(2) + ' m';
```

**Comment by pjcozzi** [comment]
> Any idea why this isn't 11.3?

Or does this have something to do with the 3D tileset?

**Comment by pjcozzi** [comment]
For the composite example, is the instance and batched tile in the composite tile?  Looks like we need to move the debug color selection logic into the composite:

image

File: Apps/Sandcastle/gallery/3D Tiles.html [hunk 4]
```diff
@@ -29,35 +29,57 @@
...
+scene.fog.enabled = false;
 scene.debugShowFramesPerSecond = true;
-//viewer.extend(Cesium.viewerCesiumInspectorMixin);
+
+var cityUrl = 'http://localhost:8080/Specs/Data/Cesium3DTiles/Tilesets/Tileset/';
+var batchedUrl = 'http://localhost:8080/Specs/Data/Cesium3DTiles/Batched/BatchedWithBatchTable/';
+var instancedUrl = 'http://localhost:8080/Specs/Data/Cesium3DTiles/Instanced/InstancedWithBatchTable/';
+var compositeUrl = 'http://localhost:8080/Specs/Data/Cesium3DTiles/Composite/Composite/';
+var pointsUrl = 'http://localhost:8080/Specs/Data/Cesium3DTiles/Points/Points/';
+var washingtonDCUrl = 'http://localhost:8002/tilesets/WashingtonDC/';
```
Comment (pjcozzi) [inline Apps/Sandcastle/gallery/3D Tiles.html]: Let's remove this and Canary Wharf and Seattle for now.

**Comment by pjcozzi** [comment]
Just those comments.  This is great; it will be soooooo useful for testing!

**Comment by lilleyse** [comment]
> Any idea why this isn't 11.3?

Not sure.. the console prints the correct values. I'll have to look around some more.

**Comment by pjcozzi** [comment]
> This didn't require any updates to server.js? I would expect that we should add something like...

Working on this now.  Will open a PR into this branch shortly.

**Comment by lilleyse** [comment]
This is updated. I changed the organization for debug coloring so that most of the work is now in `Cesium3DTile` and each content provider implements their method of handling the debug update.

 I'm still not sure why the labeling is weird, I've tried different string combinations and sometimes it works fine, sometimes nothing.

**Comment by pjcozzi** [comment]
Do the `palette`, `Change id style`, and `Change Height style` options still work for you?

**Comment by pjcozzi** [comment]
Just that question.  This looks good!

**Comment by lilleyse** [comment]
Result of a bad merge, fixed now.

**Comment by pjcozzi** [comment]
Looks good!

File: Apps/Sandcastle/gallery/3D Tiles.html [hunk 5]
```diff
@@ -29,35 +29,54 @@
...
+var viewer = new Cesium.Viewer('cesiumContainer', {
+    scene3DOnly : true
 });
 
+var scene = viewer.scene;
+scene.fog.enabled = false;
 scene.debugShowFramesPerSecond = true;
-//viewer.extend(Cesium.viewerCesiumInspectorMixin);
+
+var tilesetUrl = 'http://localhost:8080/Specs/Data/Cesium3DTiles/Tilesets/Tileset/';
```
Comment (mramato) [inline Apps/Sandcastle/gallery/3D Tiles.html]: These urls should not be hardcoded and they should not be absolutely.  They need to be relative in order to work in all cases and on the Cesium website.

Comment (emackey) [inline Apps/Sandcastle/gallery/3D Tiles.html]: Yes, this is important.  Please use a relative path.

**Pull request merged**